packages/core/root/core-root-server-internal/src/server.ts

```diff
@@ -364,6 +369,13 @@ export class Server {
     await this.resolveSavedObjectsStartPromise!(savedObjectsStart);
 
     soStartSpan?.end();
+
+    if (this.nodeRoles?.migrator === true) {
+      startTransaction?.end();
+      this.log.info('Detected migrator node role; shutting down Kibana...');
+      throw new CriticalError('Migrations completed, shutting down Kibana', 'MigrationOnlyNode', 0);
```

Comment:
We use a special error `MigrationOnlyNode` since it does most of what we need except for the "fatal" and "error" logging. I have not seen something like this done elsewhere, but a different approach might need a bit more sophistication.